text/0000-custom-components.md

```diff
@@ -0,0 +1,1113 @@
+```
+
+## Recycling Components
+
+This example implements an API which maintain a pool of recycled component
+instances to avoid allocation costs, similar to Flex's "sustain" feature.
+
+This example also make use of the "state bucket" pattern.
+
+```js
+// ember-component-pool/app/component-managers/pooled.js
+
+import { capabilities } from '@ember/component';
+import EmberObject from '@ember/object';
+
+// How many instances to keep (per type/factory)
+const LIMIT = 10;
+
+export default EmberObject.extend({
+  capabilities: capabilities('3.2', {
+    destructor: true
+  }),
+
+  init() {
+    this.pool = new Map();
+  },
+
+  createComponent(factory, args) {
+    let instances = this.pool.get(factory);
+    let instance;
+
+    if (instances && instances.length > 0) {
+      instance = instances.pop();
+      instance.setProperties(args.named);
+    } else {
+      instance = factroy.create(args.named);
+    }
+
+    // We need to remember which factory does the instance belong to so we can
+    // check it back into the pool later.
+    return { factory, instance };
+  },
+
+  getContext({ instance }) {
+    return instance;
+  },
+
+  updateComponent({ instance }, args) {
+    instance.setProperties(args.named);
+  },
+
+  destroyComponent({ factory, instance }) {
+    let instances;
+
+    if (this.pool.has(factory)) {
+      instances = this.pool.get(factory);
+    } else {
+      this.pool.set(factory, instances = []);
+    }
+
+    if (instances.length >= LIMIT) {
+      instance.destroy();
+    } else {
+      // User hook to reset any state
+      instance.willRecycle();
+      instances.push(instance);
+    }
+  },
+
+  // This is the `Ember.Object` lifecycle method, called when the component
+  // manager instance _itself_ is being destroyed, not to be confused with
+  // `destroyComponent`
+  willDestroy() {
+    for (let instances of this.pool.values()) {
+      instances.forEach(instance => instance.destroy());
+    }
+
+    this.pool.clear();
+  }
+});
+```
+
+```js
+// ember-component-pool/addon/index.js
+
+import EmberObject from '@ember/object';
+import { setComponentManager } from '@ember/component';
+
+function NOOP() {}
+
+export default setComponentManager(EmberObject.extend({
+  // Override this to implement reset any state on the instance
+  willRecycle(): NOOP,
+
+  // ...
+}), 'pooled');
+```
+
+# How We Teach This
+
+What is proposed in this RFC is a *low-level* primitive. We do not expect most
+users to interact with this layer directly. Instead, most users will simply
+benefit from this feature by subclassing these special components provided by
+addons.
+
+At present, the classic components APIs is still the primary, recommended path
+for almost all use cases. This is the API that we should teach new users, so we
+do not expect the guides need to be updated for this feature (at least not the
+components section).
+
+For documentation purposes, each Ember.js release will only document the latest
```

Comment:
I don't think this is good enough. We need to maintain a canonical list of capabilities that are supported (and any options that were allowed with each).